File: trunk/proj/Android/src/com/jasonzqshen/familyaccounting/core/listeners/ListenersManagement.java
package com.jasonzqshen.familyaccounting.core.listeners;

import java.util.ArrayList;

import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataBase;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataFactoryBase;
import com.jasonzqshen.familyaccounting.core.transaction.HeadEntity;

public class ListenersManagement {
	private final ArrayList<LoadDocumentListener> _loadDocumentListeners;
	private final ArrayList<LoadMasterDataListener> _loadMasterDataListeners;
	private final ArrayList<CreateMasterDataListener> _createMasterDataListeners;
	private final ArrayList<SaveDocumentListener> _saveDocListeners;

	public ListenersManagement() {
		_loadDocumentListeners = new ArrayList<LoadDocumentListener>();
		_loadMasterDataListeners = new ArrayList<LoadMasterDataListener>();
		_createMasterDataListeners = new ArrayList<CreateMasterDataListener>();
		_saveDocListeners = new ArrayList<SaveDocumentListener>();
	}

	/**
	 * add load document listener
	 * 
	 * @param listener
	 *            load document listener
	 */
	public void addLoadDocListener(LoadDocumentListener listener) {
		_loadDocumentListeners.add(listener);
	}

	/**
	 * add load document listener
	 * 
	 * @param listener
	 *            load document listener
	 */
	public void addLoadMasterListener(LoadMasterDataListener listener) {
		_loadMasterDataListeners.add(listener);
	}

	/**
	 * add load document listener
	 * 
	 * @param listener
	 *            load document listener
	 */
	public void addSaveDocListener(SaveDocumentListener listener) {
		_saveDocListeners.add(listener);
	}

	/**
	 * add load document listener
	 * 
	 * @param listener
	 *            load document listener
	 */
	public void addCreateMasterListener(CreateMasterDataListener listener) {
		_createMasterDataListeners.add(listener);
	}

	/**
	 * raise save documents successfully
	 * 
	 * @param doc
	 */
	public void saveDoc(HeadEntity doc) {
		for (SaveDocumentListener l : _saveDocListeners) {
			l.onSaveDocumentListener(doc);
		}
	}

	/**
	 * raise load document from disk successfully
	 * 
	 * @param source
	 *            the object to invoke the loading
	 * @param doc
	 */
	public void loadDoc(Object source, HeadEntity doc) {
		for (LoadDocumentListener l : _loadDocumentListeners) {
			l.onLoadDocumentListener(source, doc);
		}
	}

	/**
	 * raise created master data successfully
	 * 
	 * @param masterData
	 */
	public void createMasterData(MasterDataFactoryBase factory,
			MasterDataBase masterData) {
		for (CreateMasterDataListener l : _createMasterDataListeners) {
			l.onCreateMasterDataListener(factory, masterData);
		}
	}

	/**
	 * raise save master data successfully
	 * 
	 * @param masterData
	 */
	public void loadMasterData(Object source, MasterDataBase masterData) {
		for (LoadMasterDataListener l : _loadMasterDataListeners) {
			l.onLoadMasterDataListener(source, masterData);
		}
	}
}

File: branches/V1.0/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/Listeners/ListenersManagement.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WinStore_FamilyAccountingCore.MasterData;
using WinStore_FamilyAccountingCore.Transaction;
using WinStore_FamilyAccountingCore.Utilities;

namespace WinStore_FamilyAccountingCore.Listeners
{
    public delegate void DocumentLoad(Object source, HeadEntity document);
    public delegate void MasterDataLoad(Object source, MasterDataBase masterData);
    public delegate void DocumentSaved(HeadEntity document);
    public delegate void MasterDataCreated(MasterDataFactoryBase factory, MasterDataBase master);
    public delegate void DocumentReversed(HeadEntity doc);
    public delegate void GLAccountInitAmountChanged(MasterDataIdentity_GLAccount glAccountId
        , CurrencyAmount orgAmount, CurrencyAmount newAmount);

    /// <summary>
    /// List listener management
    /// </summary>
    public class ListenersManagement
    {
        public event DocumentLoad DocumentLoadHandler;
        public event MasterDataLoad MasterDataLoadHandler;
        public event DocumentSaved DocumentSavedHandler;
        public event MasterDataCreated MasterDataCreatedHandler;
        public event DocumentReversed DocumentReversedHandler;
        public event GLAccountInitAmountChanged GLAccountInitAmountChangedHandler;

        /// <summary>
        /// Construct
        /// </summary>
        public ListenersManagement()
        {

        }

        /// <summary>
        /// Raise save document event
        /// </summary>
        /// <param name="doc"></param>
        public void SaveDoc(HeadEntity doc)
        {
            foreach (DocumentSaved d in DocumentSavedHandler.GetInvocationList())
            {
                d(doc);
            }
        }

        /// <summary>
        /// Raise reverse document event
        /// </summary>
        /// <param name="doc"></param>
        public void ReverseDoc(HeadEntity doc)
        {
            foreach (DocumentReversed d in DocumentReversedHandler.GetInvocationList())
            {
                d(doc);
            }
        }

        /// <summary>
        /// Raise load document event
        /// </summary>
        /// <param name="source"></param>
        /// <param name="doc"></param>
        public void LoadDoc(Object source, HeadEntity doc)
        {
            foreach (DocumentLoad d in DocumentLoadHandler.GetInvocationList())
            {
                d(source, doc);
            }
        }

        /// <summary>
        /// Raise master data created event
        /// </summary>
        /// <param name="factory"></param>
        /// <param name="masterData"></param>
        public void CreateMasterData(MasterDataFactoryBase factory,
                MasterDataBase masterData)
        {
            foreach (MasterDataCreated d in MasterDataCreatedHandler.GetInvocationList())
            {
                d(factory, masterData);
            }
        }

        /// <summary>
        /// Master data load
        /// </summary>
        /// <param name="source"></param>
        /// <param name="masterData"></param>
        public void LoadMasterData(Object source, MasterDataBase masterData)
        {
            foreach (MasterDataLoad d in MasterDataLoadHandler.GetInvocationList())
            {
                d(source, masterData);
            }
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="source"></param>
        /// <param name="glAccountId"></param>
        /// <param name="orgAmount"></param>
        /// <param name="newAmount"></param>
        public void GLAccountInitAmountChanged(Object source, MasterDataIdentity_GLAccount glAccountId,
            CurrencyAmount orgAmount, CurrencyAmount newAmount)
        {
            foreach (GLAccountInitAmountChanged d in 
                GLAccountInitAmountChangedHandler.GetInvocationList())
            {
                d(glAccountId, orgAmount, newAmount);
            }
        }
    }
}
